test/test_replace.py

```diff
@@ -0,0 +1,239 @@
+        ),
+        (CD, "", {}),
+        pytest.param(
+            lambda: "str_obj",
+            "",
+            {},
+            marks=pytest.mark.xfail(reason="Raise TypeError if obj is not dataclass instances"),
+        ),
+        (
+            OuterDataclass,
+            "--nested_arg nested_arg_2 --arg1 1 --arg2 2.0",
+            {"nested": {"nested_arg": "nested_arg_2", "arg1": 1, "arg2": 2.0}},
+        ),
+    ],
+)
+def test_replace_nested_dict(config_cls: type, args: str, changes: dict):
+    config = config_cls()
+    config_replaced = replace(config, changes)
+    logger.info(config_replaced)
+    logger.info(config.setup(args))
+    assert config.setup(args) == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@pytest.mark.parametrize(
+    ("config_cls", "args", "changes"),
+    [
+        (CD, "--c_or_d d --d 1", {"__subgroups__@c_or_d": "d", "c_or_d.d": 1}),
+        (CD, "--c_or_d c --c True", {"__subgroups__@c_or_d": "c", "c_or_d.c": True}),
+        (
+            OuterDataclass,
+            "--some_arg some_arg_1 --nested_arg nested_arg_1",
+            {"some_arg": "some_arg_1", "nested.nested_arg": "nested_arg_1"},
+        ),
+    ],
+)
+def test_replace_flatten_dict(config_cls: type, args: str, changes: dict):
+    config = config_cls()
+    config_replaced = replace(config, changes)
+    assert config.setup(args) == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@dataclass
+class InnerClass:
+    arg1: int = 0
+    arg2: str = "foo"
+
+
+@dataclass(frozen=True)
+class OuterClass:
+    outarg: int = 1
+    nested: InnerClass = field(default_factory=InnerClass)
+
+
+def test_replace_outerdataclass():
+    changes_1 = {"outarg": 2, "nested.arg1": 1, "nested.arg2": "bar"}
+    changes_2 = {"outarg": 2, "nested": {"arg1": 1, "arg2": "bar"}}
+    c = OuterClass()
+    c1 = replace(c, changes_1)
+    c2 = replace(c, changes_2)
+    assert c1 == c2
+    assert id(c1) != id(c2)
+
+
+def test_replace_nested_1():
+    @dataclass
+    class C:
+        c: bool = False
+
+    @dataclass
+    class D:
+        d: int = 0
+
+    @dataclass(frozen=True)
+    class CD(TestSetup):
+        c_or_d: C | D = subgroups({"c": C, "d": D}, default="c")
+
+        other_arg: str = "bob"
+
+    config = CD.setup("--c_or_d c")
+    assert replace(config, {"c_or_d": {"c": True}}).c_or_d.c == True
+
+    config = CD.setup("--c_or_d d")
+    assert replace(config, {"c_or_d": {"d": 2}}).c_or_d.d == 2
+
+    config = CD.setup("--c_or_d d")
+    assert replace(config, {"c_or_d": D(d=2)}).c_or_d.d == 2
+
+    config = CD.setup("--c_or_d d")
+    assert replace(config, {"c_or_d": "d"}).c_or_d.d == 0
+
+
+@dataclass
+class Config(TestSetup):
+    ab_or_cd: AB | CD = subgroups(
+        {"ab": AB, "cd": CD},
+        default_factory=AB,
+    )
+
+
+@pytest.mark.parametrize(
+    ("config_cls", "args", "changes"),
+    [
+        (
+            Config,
+            "--ab_or_cd cd --c_or_d d",
```

Comment:
Same here. I suggest passing instances of `Config` to the test instead of command-line arguments. This will keep things a bit simpler and more explicit.